aten/src/ATen/native/quantized/cpu/ACLUtils.h

```diff
@@ -0,0 +1,257 @@
+#include <array>
+
+namespace at::native::acl_utils {
+
+using ACLDynamicQuantMatmulCacheKey = std::tuple<
+    int64_t, // M
+    bool, // FUSE_RELU
+    int64_t // NUM_THREADS
+    >;
+
+enum class ACLDynamicQuantMatmulCacheKeyIndex {
+  M,
+  FUSE_RELU,
+  NUM_THREADS,
+};
+
+struct ACLDynamicQuantMatmul {
+  arm_compute::Tensor src_s8_tensor;
+  arm_compute::Tensor src_fp32_tensor;
+  arm_compute::Tensor wei_tensor;
+  std::optional<arm_compute::Tensor> bia_tensor;
+  arm_compute::Tensor dst_tensor;
+  arm_compute::NEQuantizationLayer quant;
+  arm_compute::NEGEMMLowpMatrixMultiplyCore gemm;
+  arm_compute::NEActivationLayer acl_relu;
+  // configuration details for the ACL gemm
+  arm_compute::TensorInfo src_s8_tensor_info;
+  arm_compute::TensorInfo src_fp32_tensor_info;
+  arm_compute::TensorInfo wei_tensor_info;
+  std::optional<arm_compute::TensorInfo> bia_tensor_info;
+  arm_compute::TensorInfo dst_tensor_info;
+  arm_compute::GEMMInfo gemm_info;
+  arm_compute::ActivationLayerInfo acl_relu_info{
+      arm_compute::ActivationFunction::RELU};
+
+  // key for use in the cache
+  ACLDynamicQuantMatmulCacheKey key;
+
+  ~ACLDynamicQuantMatmul() {
+    // this will free memory allocated for the quantized src tensor since the
+    // allocation happened through ACL: src_s8_tensor.allocator()->allocate()
+    src_s8_tensor.allocator()->free();
+    // this will not free memory, it will just tell ACL that we're no longer
+    // using the pointer
+    wei_tensor.allocator()->free();
+    if (bia_tensor.has_value()) {
+      bia_tensor.value().allocator()->free();
+    }
+  }
+};
+
+} // namespace at::native::acl_utils
+
+struct PackedLinearWeightsACL : public PackedLinearWeightsOnednn {
+  using ACLDynamicQuantMatmul = at::native::acl_utils::ACLDynamicQuantMatmul;
+  using ACLDynamicQuantMatmulCacheKey =
+      at::native::acl_utils::ACLDynamicQuantMatmulCacheKey;
+  using ACLDynamicQuantMatmulCacheKeyIndex =
+      at::native::acl_utils::ACLDynamicQuantMatmulCacheKeyIndex;
+  PackedLinearWeightsACL(
+      std::unique_ptr<ideep::tensor> weight,
+      std::optional<ideep::tensor> bias,
+      at::Tensor orig_weight,
+      std::optional<at::Tensor> orig_bias)
+      : PackedLinearWeightsOnednn(
+            std::move(weight),
+            std::move(bias),
+            std::move(orig_weight),
+            std::move(orig_bias)) {
+    auto w = *(weight_.get());
+    k_ = w.get_dim(0);
+    n_ = w.get_dim(1);
+    wei_zero_point_ = orig_weight_.q_zero_point();
+    wei_scale_ = orig_weight_.q_scale();
+  }
+
+  int64_t k_;
+  int64_t n_;
+  int64_t wei_zero_point_;
+  double wei_scale_;
+
+  at::Tensor apply_dynamic(at::Tensor input, bool reduce_range = false)
+      override;
+  at::Tensor apply_dynamic_relu(at::Tensor input, bool reduce_range = false)
+      override;
+
+  std::shared_ptr<ACLDynamicQuantMatmul> get_acl_dynamic_quant_matmul(
+      const ACLDynamicQuantMatmulCacheKey& key) {
+    // We're only maintaining a 2 element LRU cache
```

Comment:
Sorry for nitpicks. Few thoughts:
- LRU cache idea does not seem to be unique to ACL, so implementation should exist someplace else. If not, please add the implementation to say `c10/utils/lru_cache.h` and then use it here
- Can variable name be shorter here? (just `cache` or `quant_cache`?
- Again, naming convention, as variable is private, shouldn't its name end with `_`

Reply:
LRU cache is not unique to ACL indeed. I could not find an implementation to use and given that our LRU cache impl just keeps track of two elements only, I don't see the point of making it global to PyTorch outside `ACLUtils.h`
If we end up implementing a (real) more complex LRU cache in the future, we'll add it to `c10/utils/lru_cache.h` and use it from there.

I agree with your comments about the name, it is now `cache_` in the new ghstack PR, please see [this line](https://github.com/pytorch/pytorch/pull/148585/files#diff-d0757cbe83e4fb0e7212a412a9908232ea248c54eb8b84a55c41d978750aa0f5R171)